kubernetes-sigs/vsphere-csi-driver
# Shallow recrypt block volumes created from snapshots

**What this PR does / why we need it**:
Apply a shallow recrypt spec when creating a CNS volume from existing snapshot.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
N/A

**Testing done**:
```shell
GINKGO_FOCUS=encryption make test-e2e
```
Verify PVC from snapshot is recrypted with the same EncryptionClass
  Steps:
  1. Generate encryption key
  2. Create EncryptionClass with encryption key [1]
  3. Create PVC with EncryptionClass [2]
  4. Create a dynamic volume snapshot from PVC [3]
  5. Create PVC with EncryptionClass [2] from snapshot [4]

Verify PVC from snapshot is recrypted with different EncryptionClass
  Steps:
  1. Generate encryption key
  2. Create EncryptionClass with encryption key [1]
  3. Create PVC with EncryptionClass [2]
  4. Create a dynamic volume snapshot from PVC [3]
  5. Generate second encryption key
  6. Create second EncryptionClass with encryption key [5]
  7. Create PVC with EncryptionClass [6] from snapshot [4]

Verify PVC from snapshot is encrypted with EncryptionClass
  Steps:
  1. Generate encryption key
  2. Create EncryptionClass with encryption key [1]
  3. Create PVC without encryption
  4. Create a dynamic volume snapshot from PVC [3]
  5. Create PVC with EncryptionClass [2] from snapshot [4]

Verify PVC from snapshot is decrypted
  Steps:
  1. Generate encryption key
  2. Create EncryptionClass with encryption key [1]
  3. Create PVC with EncryptionClass [2]
  4. Create a dynamic volume snapshot from PVC [3]
  5. Create PVC without encryption from snapshot [4]

**Special notes for your reviewer**:
The following crypto operations are supported when creating PVC from snapshot:
- Encrypt: Source is not encrypted, target is encrypted.
- Decrypt: Source is encrypted, target is not.
- NoOp: Both source and target are encrypted with the same key.
- ShallowRecrypt: Both source and target are encrypted with different keys.

**Release note**:

```release-note
```
